core/src/main/scala/kafka/server/ForwardingManager.scala

```diff
@@ -104,21 +112,30 @@ object ForwardingManager {
 }
 
 class ForwardingManagerImpl(
-  channelManager: NodeToControllerChannelManager
-) extends ForwardingManager with Logging {
+  channelManager: NodeToControllerChannelManager,
+  metrics: Metrics
+) extends ForwardingManager with AutoCloseable with Logging {
+
+  val forwardingManagerMetrics: ForwardingManagerMetrics = ForwardingManagerMetrics(metrics)
 
   override def forwardRequest(
     requestContext: RequestContext,
     requestBufferCopy: ByteBuffer,
+    requestCreationNs: Long,
     requestBody: AbstractRequest,
     requestToString: () => String,
     responseCallback: Option[AbstractResponse] => Unit
   ): Unit = {
     val envelopeRequest = ForwardingManager.buildEnvelopeRequest(requestContext, requestBufferCopy)
+    val requestCreationTimeMs = TimeUnit.NANOSECONDS.toMillis(requestCreationNs)
 
     class ForwardingResponseHandler extends ControllerRequestCompletionHandler {
       override def onComplete(clientResponse: ClientResponse): Unit = {
 
+        forwardingManagerMetrics.queueLength.getAndDecrement()
```

Comment:
This doesn't seem to be decremented on timeout. Also, we need to decide what to do about the other metrics on timeout. Probably just record how long we waited.

Reply:
Yeah we should decrement on timeout. I think we would only want to record `QueueTimeMs` upon a timeout though, not `RemoteTimeMs` since the `onTimeout` method is only called for requests that don't make it to the controller and timeout while still on the `requestQueue`.

Reply:
ok